### PR TITLE
Rename & reimport murmurhash sources from smhasher repo

### PR DESCRIPTION
**Pull request by georgthegreat**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: contrib/murmurhash/src/MurmurHash3.cpp [hunk 1]
```diff
@@ -54,9 +54,7 @@ inline uint64_t rotl64 ( uint64_t x, int8_t r )
 
 FORCE_INLINE uint32_t getblock32 ( const uint32_t * p, int i )
 {
-  uint32_t res;
-  memcpy(&res, p + i, sizeof(res));
-  return res;
+  return p[i];
```
Comment (georgthegreat) [inline contrib/murmurhash/src/MurmurHash3.cpp:57]: Looks like ub patch to me.
@alexey-milovidov, could you confirm?

File: contrib/murmurhash/include/MurmurHash3.h [hunk 2]
```diff
@@ -23,10 +24,6 @@ typedef unsigned __int64 uint64_t;
 
 #endif // !defined(_MSC_VER)
 
-#ifdef __cplusplus
```
Comment (georgthegreat) [inline contrib/murmurhash/include/MurmurHash3.h:26]: I could not understand how this was intended to work.
MurmurHash3.cpp uses C++ names mangling.

File: contrib/murmurhash/src/MurmurHash3.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply contrib/murmurhash/src/MurmurHash3.cpp:57]: Yes.

File: contrib/murmurhash/include/MurmurHash3.h (hunk 2 above)
Comment (alexey-milovidov) [reply contrib/murmurhash/include/MurmurHash3.h:26]: See https://github.com/ClickHouse/ClickHouse/pull/8792

Maybe it's enough to specify `extern "C"` in declaration.

Comment (georgthegreat) [reply contrib/murmurhash/include/MurmurHash3.h:26]: I still do not understand.
When MurmurHash3 will be compiled, these functions will be mangled in C++-style, not in C.

Comment (alexey-milovidov) [reply contrib/murmurhash/include/MurmurHash3.h:26]:
```
milovidov@milovidov-desktop:~$ clang++ -xc++ -S -o test.S - <<< 'extern "C" void f(); void f() {}' && cat test.S
        .text
        .file   "-"
        .globl  f                               # -- Begin function f
        .p2align        4, 0x90
        .type   f,@function
f:                                      # @f
        .cfi_startproc
# %bb.0:
        pushq   %rbp
        .cfi_def_cfa_offset 16
        .cfi_offset %rbp, -16
        movq    %rsp, %rbp
        .cfi_def_cfa_register %rbp
        popq    %rbp
        .cfi_def_cfa %rsp, 8
        retq
.Lfunc_end0:
        .size   f, .Lfunc_end0-f
        .cfi_endproc
                                        # -- End function
        .ident  "clang version 11.0.0 (https://github.com/llvm/llvm-project.git f5df584a5077e726ad851ccfe8496deda3e5ef07)"
        .section        ".note.GNU-stack","",@progbits
        .addrsig
```

```
milovidov@milovidov-desktop:~$ clang++ -xc++ -S -o test.S - <<< 'void f(); void f() {}' && cat test.S
        .text
        .file   "-"
        .globl  _Z1fv                           # -- Begin function _Z1fv
        .p2align        4, 0x90
        .type   _Z1fv,@function
_Z1fv:                                  # @_Z1fv
        .cfi_startproc
# %bb.0:
        pushq   %rbp
        .cfi_def_cfa_offset 16
        .cfi_offset %rbp, -16
        movq    %rsp, %rbp
        .cfi_def_cfa_register %rbp
        popq    %rbp
        .cfi_def_cfa %rsp, 8
        retq
.Lfunc_end0:
        .size   _Z1fv, .Lfunc_end0-_Z1fv
        .cfi_endproc
                                        # -- End function
        .ident  "clang version 11.0.0 (https://github.com/llvm/llvm-project.git f5df584a5077e726ad851ccfe8496deda3e5ef07)"
        .section        ".note.GNU-stack","",@progbits
        .addrsig
```

**Comment by alexey-milovidov** [comment]
clang-tidy:

```
2021-06-17 20:23:59 /ClickHouse/contrib/murmurhash/include/MurmurHash2.h:6:9: error: macro name is a reserved identifier [-Werror,-Wreserved-id-macro]
2021-06-17 20:23:59 #define _MURMURHASH2_H_
```

File: contrib/murmurhash/include/MurmurHash3.h (hunk 2 above)
Comment (georgthegreat) [reply contrib/murmurhash/include/MurmurHash3.h:26]: It turns out that CGO can not parse `exter "C" void f();` syntax.

I had to use `extern "C" {}` sections instead.

File: contrib/murmurhash/src/MurmurHash3.cpp (hunk 1 above)
Comment (georgthegreat) [reply contrib/murmurhash/src/MurmurHash3.cpp]: I have restored this fix.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**